### PR TITLE
ARROW-10057: [C++] Add hand-written Parquet nested tests

### PR DESCRIPTION
**Pull request by pitrou**
One of the tests fails and is disabled.

### CODE REVIEW

**Comment by emkornfield** [comment]
LGTM, I'll look into the failing test case.

**Comment by emkornfield** [comment]
+1 thanks.

File: cpp/src/parquet/arrow/arrow_reader_writer_test.cc [hunk 1]
```diff
@@ -2434,6 +2435,145 @@ TEST(TestArrowReadWrite, CanonicalNestedRoundTrip) {
...
+      list(field("item",
+                 struct_({field("a", ::arrow::int16(), /*nullable=*/false),
+                          field("b", list(::arrow::int64()), /*nullable=*/false)}),
+                 /*nullable=*/false));
+
+  const char* json = R"([
+      [{"a": 123, "b": [1, 2, 3]}],
+      null,
+      [],
+      [{"a": 456}, {"a": 789, "b": [null]}, {"a": 876, "b": [4, 5, 6]}]])";
```
Comment (emkornfield) [inline cpp/src/parquet/arrow/arrow_reader_writer_test.cc:2487]: I think the problem with this test, was we mark B as not null in the schema, but the samebelow on the last list implicitly sets b to null for 456.  When I add "a": 456, b: [].  The test passes.

Comment (pitrou) [reply cpp/src/parquet/arrow/arrow_reader_writer_test.cc:2487]: Ah, you may be right indeed.

**Pull request closed without merging**